modules/javafx.controls/src/main/java/javafx/scene/control/skin/TextFieldSkin.java

```diff
@@ -802,25 +802,41 @@ private void updateTextPos() {
                 // empty), then we align the Text node so that the caret will
                 // appear at the left of the centered prompt.
                 newX = midPoint - promptNode.getLayoutBounds().getWidth() / 2;
-                promptNode.setLayoutX(newX);
+                if (newX > 0) {
+                    promptNode.setLayoutX(newX);
+                } else {
+                    // Align to left when prompt text length is more than text field width
+                    promptNode.setLayoutX(caretWidth / 2);
+                }
             } else {
                 newX = midPoint - textNodeWidth / 2;
-            }
-            // Update if there is space on the right
-            if (newX + textNodeWidth <= textRight.get()) {
-                textTranslateX.set(newX);
+                // Update if there is space on the right
+                if (newX + textNodeWidth <= textRight.get()) {
+                    textTranslateX.set(newX);
+                } else if (newX < 0 && oldX > 1) {
+                    textTranslateX.set(caretWidth / 2);
+                }
             }
             break;
 
           case RIGHT:
-            newX = textRight.get() - textNodeWidth - caretWidth / 2;
-            // Update if there is space on the right
-            if (newX > oldX || newX > 0) {
-                textTranslateX.set(newX);
-            }
             if (usePromptText.get()) {
-                promptNode.setLayoutX(textRight.get() - promptNode.getLayoutBounds().getWidth() -
-                                      caretWidth / 2);
+                double promptOldX = promptNode.getLayoutX();
+                double promptNewX = textRight.get() - promptNode.getLayoutBounds().getWidth() - caretWidth / 2;
+                if(promptNewX > promptOldX || promptNewX > 0) {
+                    promptNode.setLayoutX(promptNewX);
+                } else {
+                    // Align to left when prompt text length is more than text field width
+                    promptNode.setLayoutX(caretWidth / 2);
+                }
```

Comment:
Similar comment, I don't think its correct. It's just odd that promptNewX  may be smaller than caretWidth / 2 but is accepted, but when the text is too wide the "minimum" value suddently is caretWidth / 2.

Also, I don't see how `promptOldX` has any relevance when it comes to positioning the prompt.  I think this is only relevant for the actual text (because it can scroll depending on where the cursor is), not for the prompt -- unless the prompt can be scrolled as well.  If that's the case then the `CENTER` case may need to be updated to take `promptOldX` into account as well.  As it stands currently, they have behaviors that seem to conflict.

Reply:
Yes `promptOldX` is not required in case of prompt text. Code is updated.
Same comment as the first one for the condition while comparing newly calculated value with `careWidth/2`.